### PR TITLE
Fix List behaviour and added new feature

### PR DESCRIPTION
**Pull request by cassiobotaro**
Now List has the behavior equal your comments(look at __getitem__) and we can do:
test = request.args(0 , otherwise=URL('default', 'index')) and i will be redirected if args is not found.
Otherwise parameter is not binded with cast.

### CODE REVIEW

File: gluon/storage.py [hunk 1]
```diff
@@ -269,54 +269,52 @@ def update(self, *args, **kwargs):
...
-        n = len(self)
-        if 0 <= i < n or -n <= i < 0:
-            value = self[i]
-        elif default is DEFAULT:
-            value = None
-        else:
-            value, cast = default, False
-        if cast:
-            try:
+        value = self[i] or default
```
Comment (mdipierro) [inline gluon/storage.py]: This breaks if self[i] is out of bounds.

File: gluon/storage.py [hunk 2]
```diff
@@ -269,54 +269,52 @@ def update(self, *args, **kwargs):
...
-                from http import HTTP, redirect
-                if otherwise is None:
-                    raise HTTP(404)
-                elif isinstance(otherwise, str):
-                    redirect(otherwise)
-                elif callable(otherwise):
-                    return otherwise()
-                else:
-                    raise RuntimeError("invalid otherwise")
+            if not value and otherwise:
```
Comment (mdipierro) [inline gluon/storage.py]: The ability to specify cast=int and default=None was a feature, not a bug.

**Comment by mdipierro** [comment]
It is not clear to me why the need to the change. It seems to break a lot of use cases.

File: gluon/storage.py (hunk 1 above)
Comment (cassiobotaro) [inline gluon/storage.py]: Don't break because of modification of __getitem__ function , see line 312

**Comment by cassiobotaro** [comment]
Don't break and to prove that as you can see, i added more tests. I preserve the old behaviour, but now if you can do somethink like request.args(0, otherwise=redirect(URL('default', 'index'))).
Previously the otherwise only will be considered if you have cast, now both works.

**Comment by cassiobotaro** [comment]
Another point is, the doc says "Like a regular python list but a[i] if i is out of bounds returns None
-    instead of `IndexOutOfBounds`	" but this is not True. a(i) returns None, but a[i] will raise an exception.
getitem method was rewrite to to do this now.

File: gluon/storage.py (hunk 2 above)
Comment (cassiobotaro) [inline gluon/storage.py]: I think that i don't modify it.

**Comment by mdipierro** [comment]
Before:
In [1]: print request.args(0,cast=int,default=None)
None

After:
In [1]: print request.args(0,cast=int,default=None)
HTTP: 404 NOT FOUND

This is incorrect. Will revert.

**Pull request merged**